test/jeison-test.el

```diff
@@ -218,4 +234,109 @@
+  (let ((answer-symbol (jeison-read jeison:jeison-htab-symbol
+                                    "{ \"reply\": { \"client-1a4ff5b056e2b11\" : { \"foo\": 1, \"bar\": false } } }"))
+        (answer-string (jeison-read jeison:jeison-htab-string
+                                    "{ \"reply\": { \"client-1a4ff5b056e2b11\" : { \"foo\": 1, \"bar\": false } } }")))
+    (should (and
+             (with-slots ((client-table reply)) answer-symbol
+               (let ((client-data (gethash 'client-1a4ff5b056e2b11 client-table)))
+                 (and (not (null client-data))
+                      (null (gethash 'non-existent client-table)))))
+             (with-slots ((client-table reply)) answer-string
+               (let ((client-data (gethash "client-1a4ff5b056e2b11" client-table)))
+                 (and (not (null client-data))
+                      (null (gethash "non-existent" client-table)))))))))
+
+(ert-deftest jeison:check-hash-table-of-hash-table-of-type ()
+  (jeison-defclass jeison:version-thing ()
+    ((major :initarg :major :initform 0 :type integer)
+     (minor :initarg :minor :initform 0 :type integer)))
+  (jeison-defclass jeison:jeison-thing nil
+    ((jsonFile :initarg :jsonFile :initform "invalid")
+     (kind :initarg kind :initform "nothing")
+     (version :initarg :version :type jeison:version-thing)))
+  (jeison-defclass jeison:jeison-class nil
+    ((reply :initarg :reply :type (jeison-hash-table-of symbol (jeison-hash-table-of symbol jeison:jeison-thing)))))
+  (let ((answer (jeison-read jeison:jeison-class "{
+        \"reply\" :
+        {
+            \"client-1aa66f801e65edda\" :
+            {
+                \"codemodel-v2\" :
+                {
+                    \"jsonFile\" : \"codemodel-v2-bd9606760e25788500f7.json\",
+                    \"kind\" : \"codemodel\",
+                    \"version\" :
+                    {
+                        \"major\" : 2,
+                        \"minor\" : 4
+                    }
+                }
+            },
+            \"client-1aa66f801e65beef\" :
+            {
+                \"codemodel-v2\" :
+                {
+                    \"jsonFile\" : \"codemodel-v2-bd9606760e292887f0f7.json\",
+                    \"kind\" : \"codemodel\",
+                    \"version\" :
+                    {
+                        \"major\" : 2,
+                        \"minor\" : 4
+                    }
+                }
+            }
+        }
+}")))
+    (should (with-slots ((client-table reply)) answer
+              (let ((client-1 (gethash 'client-1aa66f801e65edda client-table))
+                    (client-2 (gethash 'client-1aa66f801e65beef client-table)))
+                (and (not (null client-1))
+                     (not (null (gethash 'codemodel-v2 client-1)))
+                     (not (null client-2))
+                     (null (gethash 'non-existent client-table))))))))
+
```

Comment:
It is possible to specify hash table type for something that doesn't match our expectations. Failed type checks should also get tested for.

Reply:
Test meets my expectations. What did you really mean?

Reply:
Judging by the number of comments in this PR, the mentoring goes the other way.

Please add tests for situations when `jeison-hash-table-of` type check throws an error:

1. key type mismatch
2. value type mismatch
3. unsupported key-type, eg `jeison-hash-table-of number string`

Reply:
I think you have a misunderstanding with respect to type checking, CLOS (EIEIO) and Lisp/Elisp hash tables. You can `puthash` almost anything as a key. I can `oref` the hash table, directly `puthash` strings as keys and `gethash` will succeed, even though `jeison-read` read a class that has a hash table that uses symbols. Lisp/Elisp hash tables can have a mix of symbols and strings (and integers and classes and...) I could call `cl-check-type` directly, but that defeats the point of the test. EIEIO calls `cl-type-check` infrequently on an instance's contents.

Key and value mismatches are pointless. An unsupported key type -- you'd only indirectly alluded to key and value mismatches. Sure, I'll construct a pedantic test for an invalid key type, which tests EIEIO's type constraint system when instances are created. `jeison` doesn't emit the `signal` or `error`; EIEIO emits it.

Reply:
Or you misunderstand static typing annotations on dynamic languages and guarantees that Jeison provides.

About hash tables, we can use **exactly** the same argument for lists. List is a heterogenous collection, but at the same time `list-of TYPE` is a useful check. Why? It provides additional guarantees about what you might find in such list. Jeison uses it to check user expectations about the JSON they are trying to parse. That's it, early error to save user a headache.

Now let's go to user expectations and `jeison-hash-table-of number string`. I say that this user misunderstood how hash table parsing from JSON works and the API needs to surface an error about it as early as possible. The answer "Something will go wrong later down the line" is bad and unhelpful.

Let's move to another example:
```emacs-lisp
(jeison-defclass jeison:jeison-class nil ((x :initarg :x :type (jeison-hash-table-of string string)))
(jeison-read jeison:jeison-class "{\"x\": 42}")
```
If the user constructed this Jeison class and later receives this JSON file, we **have to** report it. Reality doesn't match the expectation. It doesn't matter if we check it in our code or if EIEIO checks it for us. It doesn't matter that the user can manually `oref` that hash table later, and `puthash` whatever they want. Type annotations is the feature of Jeison and we need to test it.

You implemented `jeison-hash-table-of` constraint, so it's not the test of EIEIO constraint system, it is a test for your code. I want to remind you that it DID HAVE A BUG. And right now there is no test covering that logic.